File: server/routers/tools.ts
import { z } from "zod";
import { router, publicProcedure } from "../_core/trpc.js";

const PYTHON_BACKEND = `http://localhost:${process.env.PYTHON_BACKEND_PORT ?? 8001}`;

async function postToPython<T>(path: string, body: Record<string, unknown>): Promise<T> {
  const response = await fetch(`${PYTHON_BACKEND}${path}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(120000),
  });
  if (!response.ok) {
    throw new Error(`Python backend error: ${response.status}`);
  }
  return response.json() as Promise<T>;
}

export const toolsRouter = router({
  ping: publicProcedure
    .input(
      z.object({
        target: z.string(),
        count: z.number().optional(),
        timeout: z.number().optional(),
      }),
    )
    .mutation(async ({ input }) => {
      return postToPython("/api/tools/ping", {
        target: input.target,
        count: input.count ?? 4,
        timeout: input.timeout ?? 5,
      });
    }),

  tcpCheck: publicProcedure
    .input(
      z.object({
        target: z.string(),
        port: z.number(),
        timeout: z.number().optional(),
      }),
    )
    .mutation(async ({ input }) => {
      return postToPython("/api/tools/tcp", {
        target: input.target,
        port: input.port,
        timeout: input.timeout ?? 5,
      });
    }),

  httpCheck: publicProcedure
    .input(
      z.object({
        url: z.string(),
        timeout: z.number().optional(),
      }),
    )
    .mutation(async ({ input }) => {
      return postToPython("/api/tools/http", {
        url: input.url,
        timeout: input.timeout ?? 10,
      });
    }),

  dnsLookup: publicProcedure
    .input(
      z.object({
        hostname: z.string(),
        recordType: z.string().optional(),
      }),
    )
    .mutation(async ({ input }) => {
      return postToPython("/api/tools/dns", {
        hostname: input.hostname,
        record_type: input.recordType ?? "A",
      });
    }),

  traceroute: publicProcedure
    .input(
      z.object({
        target: z.string(),
        maxHops: z.number().optional(),
        timeout: z.number().optional(),
      }),
    )
    .mutation(async ({ input }) => {
      return postToPython("/api/tools/traceroute", {
        target: input.target,
        max_hops: input.maxHops ?? 30,
        timeout: input.timeout ?? 5,
      });
    }),
});

File: python/routers/tools.py
"""Network diagnostic tools router"""
from fastapi import APIRouter
from pydantic import BaseModel

from services import ping_service, tcp_service, http_service, dns_service, traceroute_service

router = APIRouter(prefix="/api/tools", tags=["tools"])


# ── Request models ──────────────────────────────────────────────

class PingRequest(BaseModel):
    target: str
    count: int = 4
    timeout: int = 5


class TcpCheckRequest(BaseModel):
    target: str
    port: int
    timeout: int = 5


class HttpCheckRequest(BaseModel):
    url: str
    timeout: int = 10


class DnsLookupRequest(BaseModel):
    hostname: str
    record_type: str = "A"


class TracerouteRequest(BaseModel):
    target: str
    max_hops: int = 30
    timeout: int = 5


# ── Endpoints ───────────────────────────────────────────────────

@router.post("/ping")
def ping(req: PingRequest):
    return ping_service.execute_ping(req.target, req.count, req.timeout)


@router.post("/tcp")
def tcp_check(req: TcpCheckRequest):
    return tcp_service.execute_tcp_check(req.target, req.port, req.timeout)


@router.post("/http")
def http_check(req: HttpCheckRequest):
    return http_service.execute_http_check(req.url, req.timeout)


@router.post("/dns")
def dns_lookup(req: DnsLookupRequest):
    return dns_service.execute_dns_lookup(req.hostname, req.record_type)


@router.post("/traceroute")
def traceroute(req: TracerouteRequest):
    return traceroute_service.execute_traceroute(req.target, req.max_hops, req.timeout)
